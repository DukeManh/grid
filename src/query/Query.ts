import { IQueryAction, QueryAction } from './QueryAction';

interface IQuery {
  from: (table: string, schema?: string) => IQueryAction;
}

export class Query implements IQuery {
  from(table: string, schema?: string) {
    return new QueryAction({ name: table, schema: schema ?? 'public' });
  }
}
