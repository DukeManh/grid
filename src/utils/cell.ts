import RowService from '../services/RowService';
import { Dictionary, SupaTable } from '../types';

// TODO: should return response type obj
export async function updateCell(
  table: SupaTable,
  rowData: Dictionary<any>,
  service: RowService
): Promise<Dictionary<any> | null> {
  // find primary key
  const primaryKeys = table.columns.filter(x => x.isIdentity);
  if (!primaryKeys || primaryKeys.length == 0) return null;
  // TODO: support multi primary keys
  if (primaryKeys.length > 1) return null;

  // call RowService
  const res = await service.update(table.name, primaryKeys[0].name, rowData);
  if (res.error) return null;

  // return new rowData
  return rowData;
}
