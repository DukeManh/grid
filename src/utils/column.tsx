import * as React from 'react';
import { Column, SelectColumn, TextEditor } from 'react-data-grid';
import { Dictionary, SupaColumn, SupaTable } from '../types';
import SelectEditor from '../components/editor/SelectEditor';
import CheckboxEditor from '../components/editor/CheckboxEditor';

function setupColumnEditor(col: SupaColumn, column: Column<Dictionary<any>>) {
  if (col.isIdentity || !col.isUpdatable) return;
  const finalType = col.dataType.toLowerCase();
  console.log('finalType', finalType);
  switch (finalType) {
    case 'int4':
    case 'int8':
    case 'float4':
    case 'float8':
    case 'bigint': {
      break;
    }
    case 'boolean': {
      column.editor = CheckboxEditor;
      column.formatter = p => {
        const value = p.row[p.column.key] as boolean;
        return <>{value ? 'true' : 'false'}</>;
      };
      column.editorOptions = {
        editOnClick: true,
      };
      break;
    }
    case 'text':
    case 'varchar': {
      column.editor = TextEditor;
      break;
    }
    case 'user-defined': {
      const options = col.enums.map(x => {
        return { label: x, value: x };
      });
      column.editor = p => <SelectEditor {...p} options={options} />;
      column.editorOptions = {
        editOnClick: true,
      };
      break;
    }
    default: {
      break;
    }
  }
}

export function getGridColumns(
  table: SupaTable,
  options?: { defaultWidth?: number }
): any[] {
  const columns = table.columns.map(x => {
    let col: Column<Dictionary<any>> = {
      key: x.name,
      name: x.name,
      resizable: true,
      width: options?.defaultWidth || 100,
    };

    setupColumnEditor(x, col);
    return col;
  });
  console.log('columns', columns);
  return [SelectColumn, ...columns];
}
