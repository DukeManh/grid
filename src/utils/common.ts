import { SupaColumn, SupaRow } from '../types';

const clientOS =
  navigator.appVersion.indexOf('Win') !== -1
    ? 'windows'
    : navigator.appVersion.indexOf('Mac') !== -1
    ? 'macos'
    : 'unknown';
export const METAKEY = clientOS === 'windows' ? 'Control' : 'Command';

export function getStorageKey(prefix: string, ref: string) {
  return `${prefix}_${ref}`;
}

export function deepClone(obj: unknown) {
  try {
    return JSON.parse(JSON.stringify(obj));
  } catch (e) {
    throw e;
  }
}

export function exportRowsToCsv(
  columns: SupaColumn[],
  rows: SupaRow[],
  separator: string = ','
) {
  const keys = columns.map(x => x.name) || [];
  const csv =
    keys.join(separator) +
    '\n' +
    rows
      .map(row => {
        return keys
          .map(k => {
            let cell = row[k] === null || row[k] === undefined ? '' : row[k];
            cell =
              cell instanceof Date
                ? cell.toLocaleString()
                : cell.toString().replace(/"/g, '""');
            if (cell.search(/("|,|\n)/g) >= 0) {
              cell = `"${cell}"`;
            }
            return cell;
          })
          .join(separator);
      })
      .join('\n');
  return csv;
}
