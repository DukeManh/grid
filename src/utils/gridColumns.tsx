import * as React from 'react';
import { Column, FormatterProps } from '@supabase/react-data-grid';
import { ColumnType, SupaColumn, SupaRow, SupaTable } from '../types';
import {
  CheckboxEditor,
  DateEditor,
  DateTimeEditor,
  JsonEditor,
  ForeignKeyEditor,
  NumberEditor,
  SelectEditor,
  TextEditor,
  TimeEditor,
} from '../components/editor';
import { ColumnHeader, SelectColumn } from '../components/grid';
import { COLUMN_MIN_WIDTH } from '../constants';

export function getGridColumns(
  table: SupaTable,
  options?: {
    onEditRow?: (row: SupaRow) => void;
    defaultWidth?: string | number;
  }
): any[] {
  const selectColumn = SelectColumn(options?.onEditRow);
  const columns = table.columns.map(x => {
    let columnDef: Column<SupaRow> = {
      key: x.name,
      name: x.name,
      resizable: true,
      width: options?.defaultWidth || _getColumnWidth(x),
      minWidth: COLUMN_MIN_WIDTH,
      frozen: x.isIdentity,
    };
    const columnType = _getColumnType(x);

    columnDef.headerRenderer = props => {
      return <ColumnHeader {...props} columnType={columnType} />;
    };

    _setupColumnEditor(x, columnType, columnDef);
    return columnDef;
  });
  console.log('columns', columns);
  return [selectColumn, ...columns];
}

const DefaultFormatter = (
  p: React.PropsWithChildren<FormatterProps<SupaRow, unknown>>
) => {
  let value = p.row[p.column.key];
  if (!value) return <></>;
  if (typeof value == 'object' || Array.isArray(value)) {
    value = JSON.stringify(value);
  }
  return <>{value}</>;
};

function _setupColumnEditor(
  columnDef: SupaColumn,
  columnType: ColumnType,
  config: Column<SupaRow>
) {
  if (columnDef.isIdentity || !columnDef.isUpdatable) {
    config.formatter = DefaultFormatter;
    return;
  }

  switch (columnType) {
    case 'boolean': {
      config.editor = CheckboxEditor;
      config.formatter = p => {
        const value = p.row[p.column.key] as boolean;
        return <>{value ? 'true' : 'false'}</>;
      };
      config.editorOptions = {
        editOnClick: true,
      };
      break;
    }
    case 'date': {
      config.editor = DateEditor;
      config.formatter = DefaultFormatter;
      config.editorOptions = {
        editOnClick: true,
      };
      break;
    }
    case 'datetime': {
      config.editor = DateTimeEditor;
      config.formatter = DefaultFormatter;
      config.editorOptions = {
        editOnClick: true,
      };
      break;
    }
    case 'time': {
      config.editor = TimeEditor;
      config.formatter = DefaultFormatter;
      config.editorOptions = {
        editOnClick: true,
      };
      break;
    }
    case 'enum': {
      const options = columnDef.enum!.map(x => {
        return { label: x, value: x };
      });
      config.formatter = DefaultFormatter;
      config.editor = p => <SelectEditor {...p} options={options} />;
      config.editorOptions = {
        editOnClick: true,
      };
      break;
    }
    case 'foreign_key': {
      config.editor = ForeignKeyEditor;
      config.formatter = DefaultFormatter;
      config.editorOptions = {
        editOnClick: true,
      };
      break;
    }
    case 'json': {
      config.editor = JsonEditor;
      config.formatter = DefaultFormatter;
      config.editorOptions = {
        editOnClick: true,
      };
      break;
    }
    case 'number': {
      config.editor = NumberEditor;
      config.editorOptions = {
        editOnClick: true,
      };
      break;
    }
    case 'text': {
      config.editor = TextEditor;
      config.editorOptions = {
        editOnClick: true,
      };
      config.formatter = DefaultFormatter;
      break;
    }
    default: {
      config.formatter = DefaultFormatter;
      break;
    }
  }
}

function _getColumnType(columnDef: SupaColumn): ColumnType {
  if (columnDef.isIdentity) {
    return 'primary_key';
  } else if (_isForeignKeyColumn(columnDef)) {
    return 'foreign_key';
  } else if (_isNumericalColumn(columnDef.dataType)) {
    return 'number';
  } else if (_isJsonColumn(columnDef.dataType)) {
    return 'json';
  } else if (_isTextColumn(columnDef.dataType)) {
    return 'text';
  } else if (_isDateColumn(columnDef.format)) {
    return 'date';
  } else if (_isTimeColumn(columnDef.format)) {
    return 'time';
  } else if (_isDateTimeColumn(columnDef.format)) {
    return 'datetime';
  } else if (_isBoolColumn(columnDef.dataType)) {
    return 'boolean';
  } else if (_isEnumColumn(columnDef.dataType)) {
    return 'enum';
  } else return 'unknown';
}

function _getColumnWidth(columnDef: SupaColumn): string | number | undefined {
  if (_isNumericalColumn(columnDef.dataType)) {
    return 100;
  } else if (
    _isDateTimeColumn(columnDef.format) ||
    _isDateColumn(columnDef.format) ||
    _isTimeColumn(columnDef.format)
  ) {
    return 150;
  } else if (_isBoolColumn(columnDef.dataType)) {
    return 100;
  } else if (_isEnumColumn(columnDef.dataType)) {
    return 150;
  } else return 250;
}

const NUMERICAL_TYPES = [
  'bigint',
  'bigserial',
  'int2',
  'int4',
  'int8',
  'integer',
  'float4',
  'float8',
  'real',
  'smallint',
  'smallserial',
  'serial2',
  'serial4',
  'serial8',
];
function _isNumericalColumn(type: string) {
  return NUMERICAL_TYPES.indexOf(type.toLowerCase()) > -1;
}

const JSON_TYPES = ['json', 'jsonb'];
function _isJsonColumn(type: string) {
  return JSON_TYPES.indexOf(type.toLowerCase()) > -1;
}

const TEXT_TYPES = ['text', 'varchar'];
function _isTextColumn(type: string) {
  return TEXT_TYPES.indexOf(type.toLowerCase()) > -1;
}

const TIMESTAMP_TYPES = ['timestamp', 'timestamptz'];
function _isDateTimeColumn(type: string) {
  return TIMESTAMP_TYPES.indexOf(type.toLowerCase()) > -1;
}

const DATE_TYPES = ['date'];
function _isDateColumn(type: string) {
  return DATE_TYPES.indexOf(type.toLowerCase()) > -1;
}

const TIME_TYPES = ['time', 'timetz'];
function _isTimeColumn(type: string) {
  return TIME_TYPES.indexOf(type.toLowerCase()) > -1;
}

const BOOL_TYPES = ['boolean', 'bool'];
function _isBoolColumn(type: string) {
  return BOOL_TYPES.indexOf(type.toLowerCase()) > -1;
}

const ENUM_TYPES = ['user-defined'];
function _isEnumColumn(type: string) {
  return ENUM_TYPES.indexOf(type.toLowerCase()) > -1;
}

function _isForeignKeyColumn(columnDef: SupaColumn) {
  const { targetTableSchema, targetTableName, targetColumnName } = columnDef;
  return !!targetTableSchema && !!targetTableName && !!targetColumnName;
}
