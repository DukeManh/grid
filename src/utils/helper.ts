export const constrainDragAxis = (draggableStyle: any, axis = 'y') => {
  const { transform } = draggableStyle;
  let activeTransform = {};
  if (transform) {
    if (axis === 'x') {
      activeTransform = {
        transform: `translate(${transform.substring(
          transform.indexOf('(') + 1,
          transform.indexOf(',')
        )}, 0)`,
      };
    } else if (axis === 'y') {
      activeTransform = {
        transform: `translate(0, ${transform.substring(
          transform.indexOf(',') + 1,
          transform.indexOf(')')
        )})`,
      };
    }
  }
  return {
    ...draggableStyle,
    ...activeTransform,
  };
};
