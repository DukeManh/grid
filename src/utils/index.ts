export * from './common';
export * from './column';
export * from './filter';
export * from './row';
export * from './sort';
export * from './table';
