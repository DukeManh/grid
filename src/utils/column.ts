import { Column } from '@phamhieu1998/react-data-grid';
import { SavedState } from '../types';
import { deepClone } from './common';

export function cloneColumn(column: Column<any, any>) {
  const cloned = deepClone(column);
  // these properties can't be cloned. Need to manual re-set again
  cloned.editor = column.editor;
  cloned.headerRenderer = column.headerRenderer;
  return cloned;
}

export function getInitialGridColumns(
  gridColumns: Column<any, any>[],
  savedState?: SavedState
) {
  let result = gridColumns;
  if (savedState?.gridColumns) {
    result = [];
    for (let i = 0; i < savedState.gridColumns.length; i++) {
      const x = savedState.gridColumns[i];
      const found = gridColumns.find(y => y.key === x.name);
      // merge with savedState item props: width
      if (found) result.push({ ...found, width: x.width, frozen: x.frozen });
    }
    // console.log('exist grid columns', result);
    // check for newly created columns
    const newGridColumns = gridColumns.filter(x => {
      const found = savedState.gridColumns.find(y => y.key === x.name);
      return found ? false : true;
    });
    // console.log('newGridColumns', newGridColumns);
    result.concat(newGridColumns);
  }
  return result;
}
