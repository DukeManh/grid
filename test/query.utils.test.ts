import { formatLiteral } from '../src/query/Query.utils';

test('format array literal', async () => {
  const format = formatLiteral([321.23, 23, 54.6]);
  expect(format).toMatchSnapshot();
});
test('format json literal', async () => {
  const format = formatLiteral({ firstName: 'aaaa', lastName: 'bbb' });
  expect(format).toMatchSnapshot();
});
test('format complex json literal', async () => {
  const format = formatLiteral({
    firstName: 'aaaa',
    lastName: 'bbb',
    numbers: [1, 54, 6.7, 34.5],
  });
  expect(format).toMatchSnapshot();
});
