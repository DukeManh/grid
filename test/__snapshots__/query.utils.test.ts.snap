// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format array literal 1`] = `"'{321.23,23,54.6}'"`;

exports[`format complex json literal 1`] = `"'{\\"firstName\\":\\"aaaa\\",\\"lastName\\":\\"bbb\\",\\"numbers\\":[1,54,6.7,34.5]}'"`;

exports[`format json literal 1`] = `"'{\\"firstName\\":\\"aaaa\\",\\"lastName\\":\\"bbb\\"}'"`;
